Support UPF Paging request#7220

prabinakpattnaik opened the pull request:
Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

## Summary

UPF Paging request supported in SMF:
Received the message from UPF
Get the IMSI from mobilityd w.r.t. UE IP Address.
Read the SessionMap from global session store.
Send the notification to AMF for session info.

themarwhal commented on lte/gateway/c/session_manager/CMakeLists.txt:
```diff
@@ -54,7 +55,7 @@ endif ()
 include_directories("${OUTPUT_DIR}")
 include_directories("${MAGMA_ROOT}/orc8r/gateway/c/common/logging")
-
+include_directories("${MAGMA_ROOT}/lte/gateway/c/oai/lib/mobility_client")
```
We definitely do not want a dependency into `lte/gateway/c/oai/lib/mobility_client`. Let's create a mobilityd client in SessionD if we need to talk to mobilityd

prabinakpattnaik replied:
Yes Marie, I already fixed same things in magma-5G. I will add here.

prabinakpattnaik replied:
Changes done.

themarwhal commented on lte/protos/session_manager.proto:
```diff
@@ -1013,7 +1030,7 @@ message UPFPagingInfo {
     rpc SetUPFNodeState(UPFNodeState) returns (SmContextVoid) {}
     rpc SetUPFSessionsConfig(UPFSessionConfigState) returns (magma.orc8r.Void) {}
-    rpc SendPagingReuest(UPFPagingInfo) returns (magma.orc8r.Void) {}
+    rpc SendPagingRequest(UPFPagingInfo) returns (SmContextVoid) {}
```
Can we pull out this spelling mistake fix / void type change into a separate PR? Easier to approve simple changes like that.

prabinakpattnaik replied:
I fixed it here.

prabinakpattnaik committed:
Support UPF Paging request#7220

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+        const std::string& imsi = sid.id();
+        conv_enforcer_->get_event_base().runInEventBaseThread(
+            [this, imsi, fte_id, response_callback]() {
+              if (!imsi.length()) {
```
Can we pull out the inside of this lambda function passed into `conv_enforcer_->get_event_base().runInEventBaseThread(` into a separate function? I think it'll be cleaner to read that way.

prabinakpattnaik replied:
Created a new function and lambad function is defined inside a new function.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.h:
```diff
@@ -52,6 +53,14 @@ class UpfMsgHandler {
+
+  /**
+   * Paging Notification handling
+   *
```
extra line in comment

prabinakpattnaik replied:
Fixed.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+        conv_enforcer_->get_event_base().runInEventBaseThread(
+            [this, imsi, fte_id, response_callback]() {
+              if (!imsi.length()) {
+                MLOG(MERROR) << "Subscriber is NULL";
```
Let's be more descriptive here. 
Maybe something like
`get_subscriberid_from_ipv4 for IP <IP> returned an empty subscriber ID`

prabinakpattnaik replied:
Changed the error log statement.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+            [this, imsi, fte_id, response_callback]() {
+              if (!imsi.length()) {
+                MLOG(MERROR) << "Subscriber is NULL";
+                Status status(grpc::NOT_FOUND, "Sesion Not found");
```
`Session` typo

themarwhal replied:
but i think this error message can also be more descriptive. Specify the reason why it's failing.

prabinakpattnaik replied:
Fixed.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+              if (!session_it) {
+                MLOG(MERROR) << "No session found in SessionMap for IMSI "
+                             << imsi << " with teid " << fte_id;
+                Status status(grpc::NOT_FOUND, "Sesion Not found");
```
Sesion typo -> `session`

Specify in the message that the session was not found for IMSI/Teid pair.

prabinakpattnaik replied:
Yes. Corrected typo and changed status message.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,66 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+                return;
+              }
+
+              MLOG(MINFO) << "IDLE_MODE::: Session found in SendingPaging "
```
Log the session ID here.

prabinakpattnaik replied:
Added session_id in a log statement.

themarwhal reviewed (changes requested):
Change in general looks good. Left some comments.


Thank you Marie for reviewed the code. Addressed the review comments.

prabinakpattnaik committed:
Support UPF Paging request#7220

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

prabinakpattnaik committed:
Support UPF Paging request#7220

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

themarwhal commented on lte/gateway/c/session_manager/MobilitydClient.cpp:
```diff
@@ -0,0 +1,38 @@
+namespace magma {
+
+AsyncMobilitydClient::AsyncMobilitydClient(
+    std::shared_ptr<grpc::Channel> channel)
```
pls address
```
Add #include <memory> for shared_ptr<>  [build/include_what_you_use] [4]
```

prabinakpattnaik replied:
Addressed it.

themarwhal commented on lte/gateway/c/session_manager/MobilitydClient.cpp:
```diff
@@ -0,0 +1,38 @@
+    std::function<void(Status status, SubscriberID)> callback) {
+  auto local_resp = new AsyncLocalResponse<SubscriberID>(
+      std::move(callback), RESPONSE_TIMEOUT);
+  local_resp->set_response_reader(std::move(stub_->AsyncGetSubscriberIDFromIP(
```
pls address
```
 Add #include <utility> for move  [build/include_what_you_use] [4]
```

prabinakpattnaik replied:
Addressed it.

themarwhal commented on lte/gateway/c/session_manager/MobilitydClient.h:
```diff
@@ -0,0 +1,65 @@
+ public:
+  AsyncMobilitydClient();
+
+  AsyncMobilitydClient(std::shared_ptr<grpc::Channel> mobilityd_channel);
```
```suggestion
  explicit AsyncMobilitydClient(std::shared_ptr<grpc::Channel> mobilityd_channel);
```

prabinakpattnaik replied:
changes done as suggested.

themarwhal commented on lte/gateway/c/session_manager/MobilitydClient.h:
```diff
@@ -0,0 +1,65 @@
+
+ private:
+  static const uint32_t RESPONSE_TIMEOUT = 6;  // seconds
+  std::unique_ptr<MobilityService::Stub> stub_;
```
pls address
```
 Add #include <memory> for unique_ptr<>  [build/include_what_you_use] [4]
```

prabinakpattnaik replied:
Addressed it.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,75 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+      return;
+    }
+
+    // retreive session_map entry
```
please address 
```
[misspell] reported by reviewdog 🐶
"retreive" is a misspelling of "retrieve"
```

prabinakpattnaik replied:
correction done.

themarwhal commented on lte/gateway/c/session_manager/UpfMsgManageHandler.cpp:
```diff
@@ -123,4 +130,75 @@ void UpfMsgManageHandler::SetUPFSessionsConfig(
+    // Generate Paging trigget to AMF.
+    conv_enforcer_->handle_state_update_to_amf(
+        *session, magma::lte::M5GSMCause::OPERATION_SUCCESS, UE_PAGING_NOTIFY);
+    MLOG(MINFO) << "UPF Paging notificaiton forwarded to AMF of imsi:" << imsi;
```
pls address
```
[misspell] reported by reviewdog 🐶
"notificaiton" is a misspelling of "notification"
```

prabinakpattnaik replied:
Correction done.

themarwhal reviewed (changes requested):
please address some of the lint annotations
(especially the #include ones and spelling)

themarwhal commented:
@pshelar  / @koolzz do you mind reviewing the pipelined portion of this PR? (it's a one liner)

koolzz reviewed (approved):
pipelined oneliner looks good

prabinakpattnaik committed:
Support UPF Paging request#7220

Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

prabinakpattnaik commented:
> please address some of the lint annotations
> (especially the #include ones and spelling)

Addressed review comments and rebased with latest master. Thanks Marie.
